synapse/handlers/message.py

```diff
@@ -685,23 +704,40 @@ async def create_event(
 
         builder.internal_metadata.historical = historical
 
-        event, context = await self.create_new_client_event(
-            builder=builder,
-            requester=requester,
-            allow_no_prev_events=allow_no_prev_events,
-            prev_event_ids=prev_event_ids,
-            auth_event_ids=auth_event_ids,
-            state_event_ids=state_event_ids,
-            depth=depth,
-        )
+        if for_batch:
+            assert prev_event_ids is not None
+            assert state_map is not None
+            assert current_state_group is not None
+            auth_ids = self._event_auth_handler.compute_auth_events(builder, state_map)
+            event = await builder.build(
+                prev_event_ids=prev_event_ids, auth_event_ids=auth_ids, depth=depth
+            )
+            context = await self.state.compute_event_context_for_batched(
+                event, state_map, current_state_group
+            )
```

Comment:
I think we can move this into `create_new_client_event`? That way we can always call `create_new_client_event`, and so we don't have to worry about if we've done everything we need to, e.g. atm this code path isn't calling `third_party_event_rules.check_event_allowed` (which may or may not be an issue)?